src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/Marshalling/AnsiStringMarshaller.cs

```diff
@@ -9,90 +9,93 @@ namespace System.Runtime.InteropServices.Marshalling
     /// Marshaller for ANSI strings
     /// </summary>
     [CLSCompliant(false)]
-    [CustomTypeMarshaller(typeof(string), BufferSize = 0x100,
-        Features = CustomTypeMarshallerFeatures.UnmanagedResources | CustomTypeMarshallerFeatures.TwoStageMarshalling | CustomTypeMarshallerFeatures.CallerAllocatedBuffer)]
-    public unsafe ref struct AnsiStringMarshaller
+    [CustomMarshaller(typeof(string), MarshalMode.Default, typeof(AnsiStringMarshaller))]
+    [CustomMarshaller(typeof(string), MarshalMode.ManagedToUnmanagedIn, typeof(ManagedToUnmanagedIn))]
+    public static unsafe class AnsiStringMarshaller
     {
-        private byte* _nativeValue;
-        private bool _allocated;
+        /// <summary>
+        /// Convert a string to an unmanaged version.
+        /// </summary>
+        /// <param name="managed">A managed string</param>
+        /// <returns>An unmanaged string</returns>
+        public static byte* ConvertToUnmanaged(string? managed)
+            => (byte*)Marshal.StringToCoTaskMemAnsi(managed);
```

Comment:
`Marshal.StringToCoTaskMemAnsi` is single-pass and allocates memory conservatively (it can allocate 2x - 3x more than necessary). The existing `AnsiStringMarshaller` marshaller implementation and the existing built-in marshallers are two-pass and allocate exact amount of memory.

Not sure whether we care enough about this difference. Just pointing it out. Feel free to "won't fix".

Reply:
(same for Utf8 marshaller)

Reply:
Bah. It's a fair criticism and shouldn't be hard to do correctly. I'll make the change.